### PR TITLE
8294977: Convert test/jdk/java tests from ASM library to Classfile API

### PR DESCRIPTION
**Pull request by liach**
Summaries:
1. A few recommendations about updating the constant API is made at https://mail.openjdk.org/pipermail/classfile-api-dev/2023-March/000233.html and I may update this patch shall the API changes be integrated before
2. One ASM library-specific test, `LambdaAsm` is removed. Others have their code generation infrastructure upgraded from ASM to Classfile API.
3. Most tests are included in tier1, but some are not:
In `:jdk_io`: (tier2, part 2)
```
test/jdk/java/io/Serializable/records/SerialPersistentFieldsTest.java
test/jdk/java/io/Serializable/records/ProhibitedMethods.java
test/jdk/java/io/Serializable/records/BadCanonicalCtrTest.java
```
In `:jdk_instrument`: (tier 3)
```
test/jdk/java/lang/instrument/RetransformAgent.java
test/jdk/java/lang/instrument/NativeMethodPrefixAgent.java
test/jdk/java/lang/instrument/asmlib/Instrumentor.java
```

@asotona Would you mind reviewing?

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294977](https://bugs.openjdk.org/browse/JDK-8294977): Convert test/jdk/java tests from ASM library to Classfile API (**Sub-task** - P5)

### Reviewers
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13009/head:pull/13009` \
`$ git checkout pull/13009`

Update a local copy of the PR: \
`$ git checkout pull/13009` \
`$ git pull https://git.openjdk.org/jdk.git pull/13009/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13009`

View PR using the GUI difftool: \
`$ git pr show -t 13009`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13009.diff">https://git.openjdk.org/jdk/pull/13009.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13009#issuecomment-1467272690)

### CODE REVIEW

File: test/jdk/java/lang/ModuleTests/AnnotationsTest.java [hunk 1]
```diff
@@ -146,23 +143,15 @@ public void testWithModuleInfoResourceXXXX() throws IOException {
...
-        cr.accept(cv, attrs.toArray(new Attribute[0]), 0);
-
-        AnnotationVisitor annotationVisitor
-            = cv.visitAnnotation("Ljava/lang/Deprecated;", true);
-        annotationVisitor.visit("forRemoval", forRemoval);
-        annotationVisitor.visit("since", since);
-        annotationVisitor.visitEnd();
-
-        return cw.toByteArray();
+        return Classfile.parse(bytes).transform(ClassTransform.ACCEPT_ALL.andThen(ClassTransform.endHandler(clb -> {
```
Comment (AlanBateman) [inline test/jdk/java/lang/ModuleTests/AnnotationsTest.java:146]: The conversion of this test okay but would be good if you split up the overly long lines as they are inconsistent with everything else in this test and makes it annoying to look at the changes side-by-side.

File: test/jdk/java/util/ServiceLoader/BadProvidersTest.java [hunk 2]
```diff
@@ -207,55 +210,35 @@ public void testBadProvider(String testName, String ignore) throws Exception {
...
-                false);
-        mv.visitInsn(ARETURN);
-        mv.visitMaxs(0, 0);
-        mv.visitEnd();
-
-        cw.visitEnd();
+        var bytes = Classfile.build(ClassDesc.of("p", "ProviderFactory"), clb -> {
+            clb.withSuperclass(CD_Object);
+            clb.withFlags(AccessFlag.PUBLIC, AccessFlag.SUPER);
+            var provider$1Desc = ClassDesc.of("p", "ProviderFactory$1");
```
Comment (AlanBateman) [inline test/jdk/java/util/ServiceLoader/BadProvidersTest.java:216]: This is class descriptor for ProviderFactory$1, not "Provider" so maybe rename this to providerFactory1 or something a bit clearer.

File: test/jdk/java/lang/invoke/defineHiddenClass/BasicTest.java [hunk 3]
```diff
@@ -263,46 +269,17 @@ private Object[][] emptyClasses() {
...
-                throw new IllegalArgumentException("unexpected access flag: " + accessFlags);
-        }
+        assertEquals(hc.isSynthetic(), accessFlags.contains(SYNTHETIC));
+        assertEquals(hc.isEnum(), accessFlags.contains(ENUM));
+        assertEquals(hc.isAnnotation(), accessFlags.contains(ANNOTATION));
+        assertEquals(hc.isInterface(), accessFlags.contains(INTERFACE));
+
         assertTrue(hc.isHidden());
-        assertTrue(hc.getModifiers() == (ACC_PUBLIC|accessFlags));
+        assertEquals(hc.getModifiers(), accessFlags.stream().mapToInt(AccessFlag::mask).reduce(AccessFlag.PUBLIC.mask(), (a, b) -> a | b));
```
Comment (AlanBateman) [inline test/jdk/java/lang/invoke/defineHiddenClass/BasicTest.java:282]: Do you mind splitting this line too, it's 140+ characters long so impossible to look at the changes side-by-side.

**Review by AlanBateman — COMMENTED** [review]
Good to see these tests converted, just a few nits about trying to keep the code/style consistent with the existing code/style where possible.

File: test/jdk/java/lang/Class/getSimpleName/GetSimpleNameTest.java [hunk 4]
```diff
@@ -121,88 +136,91 @@ protected Class<?> findClass(String name) throws ClassNotFoundException {
...
-            cw.visitInnerClass(innerName, outerName, simpleName, ACC_PUBLIC | ACC_STATIC);
-
-            makeDefaultCtor(cw);
-            cw.visitEnd();
-            return cw.toByteArray();
+            var name = (isInner ? innerName : outerName);
+            return Classfile.build(name, clb -> {
+                clb.withSuperclass(CD_Object);
+                clb.withFlags(AccessFlag.PUBLIC, AccessFlag.SUPER);
+                clb.accept(InnerClassesAttribute.of(
```
Comment (asotona) [inline test/jdk/java/lang/Class/getSimpleName/GetSimpleNameTest.java:174]: During the API discussions there was slightly more recommended to use `ClasfileBuilder::with` over `ClasfileBuilder::accept`, however it is just a cosmetic difference.

**Comment by asotona** [comment]
I like the easy way I can read the tests code now even I don't know them.
They look great :)

File: test/jdk/java/util/ServiceLoader/BadProvidersTest.java (hunk 2 above)
Comment (AlanBateman) [reply test/jdk/java/util/ServiceLoader/BadProvidersTest.java]: > This is class descriptor for ProviderFactory$1, not "Provider" so maybe rename this to providerFactory1 or something a bit clearer.

The updated version looks good. I assume you'll do a pass over the updated tests to bump their copyright date as this is the first change in 2023 for many of these tests.

File: test/jdk/java/lang/invoke/8022701/MHIllegalAccess.java [hunk 5]
```diff
@@ -25,56 +25,53 @@
...
 
-public class MHIllegalAccess implements Opcodes {
+import static jdk.internal.classfile.Classfile.ACC_PRIVATE;
+import static jdk.internal.classfile.Classfile.ACC_PROTECTED;
+import static jdk.internal.classfile.Classfile.ACC_PUBLIC;
 
-   public static void main(String args[]) throws Throwable  {
+public class MHIllegalAccess {
+
+   public static void main(String[] args) throws Throwable  {
```
Comment (turbanoff) [inline test/jdk/java/lang/invoke/8022701/MHIllegalAccess.java:50]:
```suggestion
   public static void main(String[] args) throws Throwable {
```

**Comment by liach** [comment]
keep alive. this might not be integrated until classfile api has been previewed.

**Comment by liach** [comment]
keep-alive; this will be updated for the classdile object update.

**Comment by liach** [comment]
Should I wait for Classfile API transition to preview before I come back to this?

**Comment by asotona** [comment]
I would integrate conversions that are ready, otherwise the work spend to fix and reflect changes would double in the preview branch.

**Comment by liach** [comment]
Keeping alive, I plan to return to this patch after the integration of Class-File API transition to preview.

**Comment by liach** [comment]
I have migrated the tests to Preview Class-File API and fixed a few issues related to preview status that broke the tests. All tests now pass.

**Comment by liach** [comment]
Keep alive, @asotona can you review this again as it's up-to-date against current master's Classfile API? (i.e. without the instruction renames)

**Review by asotona — APPROVED** [review]
Yes, CodeBuilder method renames will require some refactoring.
This is good.

**Comment by jaikiran** [comment]
I see that this PR was approved a month back and there haven't been any changes to the PR since this (which is a good thing). I'll run this against our internal CI to make sure this still works as expected and will sponsor this later today, if that run goes fine.

**Comment by jaikiran** [comment]
tier1, tier2 and tier3 tests went fine with these changes on top of latest master branch.

**Pull request closed without merging**